### PR TITLE
Start enforcing Ruff rule ANN201

### PR DESCRIPTION
**Pull request by eerovaher**
### Description

[ANN201](https://docs.astral.sh/ruff/rules/missing-return-type-undocumented-public-function/) checks for public functions without return type annotation. The permanent configuration is updated to ignore the rule in tests to avoid having to clutter them with the `-> None` return type annotation. The violations outside the tests are fixed.

Note that Ruff is configured to check this rule only for functions that already have type annotations in their signature.

NOTE: When this pull request was first opened it did not configure Ruff to ignore `ANN201` in tests and instead added the `-> None` annotations to some tests in `ŧable`, but the configuration update removed the need to edit that code.

A follow-up pull request will address ANN201 in `modeling`.

### Approvals by subpackage:

- [x] `cosmology` https://github.com/astropy/astropy/pull/15685#pullrequestreview-1822165615

- [x] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/table/tests/test_groups.py
Comment (eerovaher) [inline astropy/table/tests/test_groups.py:1]: The vast majority of functions in `test_*.py` files are actual tests which are not meant to be explicitly called by users and always return `None`. It might therefore be a good idea to permanently ignore ANN201 in tests.  For now I've added the annotations here to get the discussion going.

File: astropy/modeling/bounding_box.py
Comment (neutrinoceros) [inline astropy/modeling/bounding_box.py:685]: In my experience, unions in return type annotations are sometimes worse than no annotation, because IDEs and type checkers will insist that all cases are matched in user code, even when it doesn't make sense. Here however I don't immediately see how to avoid it, as the return type doesn't depend on the arguments (excluding `TypeVar` as a solution), but on the internal state of the object (the `len` of `self._intervals`). If anyone knows any way to express that in type annotations, this would be useful here.

File: astropy/cosmology/core.py
Comment (nstarman) [inline astropy/cosmology/core.py:522]:
```suggestion
    def clone(
        self, *, meta: Mapping | None = None, to_nonflat: bool = False, **kwargs
    ) -> Self:
```

File: astropy/table/tests/test_groups.py
Comment (nstarman) [reply astropy/table/tests/test_groups.py:1]: I think that's a good idea.

Comment (nstarman) [reply astropy/table/tests/test_groups.py:1]: If the necessity arises, we can change this to a more granular per-file ignore in the tests. But for now a blanket ignore seems reasonable.

**Review by taldcroft — APPROVED** [review]
Fine by me for Table.

File: astropy/cosmology/core.py
Comment (eerovaher) [reply astropy/cosmology/core.py:522]: I thought that `Self` might not be the correct return type, depending on the value of `to_nonflat`.

Comment (nstarman) [reply astropy/cosmology/core.py:522]: It's true in detail but in general this is more useful. The most correct would be to use a typing overload.

Comment (eerovaher) [reply astropy/cosmology/core.py:522]: You know `cosmology` much better than I do, so if you think `Self` is a more useful annotation then I'll use that. I'd just like #15686 to be resolved before I push changes here.

File: astropy/cosmology/tests/test_core.py
Comment (nstarman) [inline astropy/cosmology/tests/test_core.py:32]: `np.ndarray` accepts 2 type parameters. Or `NDArray` from `numpy.typing` takes only the value type

**Review by nstarman — COMMENTED** [review]
Cosmology LGTM. One comment on `ndarray`s.

File: astropy/cosmology/tests/test_core.py
Comment (eerovaher) [reply astropy/cosmology/tests/test_core.py]: And additionally I had missed one `list` from the annotation.

**Review by nstarman — APPROVED** [review]
Thanks!

**Comment by eerovaher** [comment]
I had to update this because of #15870.

**Review by nstarman — APPROVED** [review]
Cosmology changes LGTM. Thanks!

**Comment by eerovaher** [comment]
If the changes to `modeling` don't get reviewed by Friday then I'll move them to a separate pull request so that the changes to the global Ruff settings and `cosmology` can be merged.

**Comment by eerovaher** [comment]
I had to rebase this because adding ANN201 to the temporary ignores for `modeling` would have caused a merge conflict otherwise.

**Comment by eerovaher** [comment]
@nstarman, this pull request now only edits Ruff configuration and `cosmology`, so we don't have to wait for reviews by any other sub-package maintainers anymore.

**Pull request merged**
